File: source/go/struct_of_arrays.go
package main

import (
	"fmt"
	"math"
)

// SoA layout: parallel arrays for x, y, z coordinates
const COUNT = 5

var xs [COUNT]float64
var ys [COUNT]float64
var zs [COUNT]float64

func setPoint(i int, x, y, z float64) {
	xs[i] = x
	ys[i] = y
	zs[i] = z
}

func distance(i int) float64 {
	x := xs[i]
	y := ys[i]
	z := zs[i]
	return math.Sqrt(x*x + y*y + z*z)
}

func main() {
	setPoint(0, 1.0, 0.0, 0.0)
	setPoint(1, 0.0, 2.0, 0.0)
	setPoint(2, 0.0, 0.0, 3.0)
	setPoint(3, 1.0, 1.0, 1.0)
	setPoint(4, 3.0, 4.0, 0.0)

	for i := 0; i < COUNT; i++ {
		x := xs[i]
		y := ys[i]
		z := zs[i]
		d := distance(i)
		fmt.Printf("point %d: (%g, %g, %g) distance = %g\n", i, x, y, z, d)
	}

	fmt.Println("all arrays freed")
}

File: source/javascript/struct_of_arrays.js
// SoA layout with Float64Arrays (parallel arrays for x, y, z coordinates)
const COUNT = 5;
const xs = new Float64Array(COUNT);
const ys = new Float64Array(COUNT);
const zs = new Float64Array(COUNT);

function setPoint(i, x, y, z) { xs[i] = x; ys[i] = y; zs[i] = z; }
function getX(i) { return xs[i]; }
function getY(i) { return ys[i]; }
function getZ(i) { return zs[i]; }
function distance(i) { return Math.sqrt(xs[i] ** 2 + ys[i] ** 2 + zs[i] ** 2); }

setPoint(0, 1.0, 0.0, 0.0);
setPoint(1, 0.0, 2.0, 0.0);
setPoint(2, 0.0, 0.0, 3.0);
setPoint(3, 1.0, 1.0, 1.0);
setPoint(4, 3.0, 4.0, 0.0);

for (let i = 0; i < COUNT; i++) {
    const x = getX(i), y = getY(i), z = getZ(i), d = distance(i);
    console.log(`point ${i}: (${x}, ${y}, ${z}) distance = ${d}`);
}
console.log("all arrays freed");
